README.md
# Scope

The set of monitoring plugins in this repository is intended to serve
as icinga / nagios / .. checks for DANE enabled network services. The
central functionality therefore is checking whether the services
conforms with the DANE records. However, all service plugins are also
supposed to work as *the* plugin for each service and should be able
to monitor all relevant properties.

# Getting the Code

The main repository is located at [git.siccegge.de][1] but mirrors can
be found on both [GitLab][2] and [GitHub][3]. Pull request will be
processed by mail as well as on both platforms.

[1]: https://git.siccegge.de/?p=dane-monitoring-plugins.git
[2]: https://gitlab.com/siccegge/dane-monitoring-plugins
[3]: https://github.com/siccegge/dane-monitoring-plugins

# Dependencies

The plugins need `pyasn1`, `pyasn1-modules` and the python `unbound`
as well as python3. The `check_dnssec` module needs additionally the
python `ldns` module. Unfortunately the unbound package in Debian only
provides python2 modules currently. Building unbound with python3
support from source works fine however.

# License

Unfortunately the problems at hand tend to result in a dependency on
OpenSSL which makes plain GPL inviable. Until a proper decision is
made you can assume GPLv3 with OpenSSL exception


debian/changelog
dane-monitoring-plugins (0.1) siccegge-stretch; urgency=medium

  * Initial release. (Closes: #XXXXXX)

 -- Christoph Egger <christoph@debian.org>  Sat, 06 Jan 2018 18:25:00 +0100


check_dane/cert.py
#!/usr/bin/python3

from datetime import datetime
import logging
from ssl import cert_time_to_seconds

from pyasn1_modules import rfc2459
from pyasn1.codec.der import decoder, encoder

def verify_certificate(cert, args):
    expiretimestamp = cert_time_to_seconds(cert['notAfter'])
    starttimestamp = cert_time_to_seconds(cert['notBefore'])

    if datetime.utcfromtimestamp(starttimestamp) > datetime.utcnow():
        logging.error("Certificate will only be valid starting %s", cert['notBefore'])
        return 2

    if datetime.utcfromtimestamp(expiretimestamp) < datetime.utcnow():
        logging.error("Certificate will only be valid until %s", cert['notAfter'])
        return 2

    delta = datetime.utcfromtimestamp(expiretimestamp) - datetime.utcnow()
    deltastr = str(delta).split(",")

    if delta.days < args.critdays:
        logging.error("expires in %8s,%16s", deltastr[0], deltastr[1])
        return 2
    elif delta.days < args.warndays:
        logging.warning("expires in %8s,%16s", deltastr[0], deltastr[1])
        return 1

    return 0

def get_spki(certificate):
    cert = decoder.decode(certificate, asn1Spec=rfc2459.Certificate())[0]
    spki = cert['tbsCertificate']["subjectPublicKeyInfo"]
    return encoder.encode(spki)

def add_certificate_options(argparser):
    argparser.add_argument("--warndays", type=int, default=-1,
                           help="Days before certificate expiration to warn")
    argparser.add_argument("--critdays", type=int, default=-1,
                           help="Days before certificate expiration to raise error")


check_dane/tlsa.py
#!/usr/bin/python3

import sys
import codecs
import hashlib
import logging

from .cert import get_spki

from unbound import ub_strerror

try:
    from unbound import RR_TYPE_TLSA
except ImportError:
    RR_TYPE_TLSA = 52



class TLSARecord:
    """Class representing a TLSA record"""
    def __init__(self, usage, selector, matching, payload):
        self._usage = usage
        self._selector = selector
        self._matching = matching
        self._payload = payload


    def match(self, certificate):
        """Returns true if the certificate is covered by this TLSA record"""
        if self._selector == 0:
            verifieddata = certificate
        elif self._selector == 1:
            verifieddata = get_spki(certificate)
        else:
            # currently only 0 and 1 are assigned
            sys.stderr.write("Only selectors 0 and 1 supported\n")

        if self._matching == 0:
            if verifieddata == self._payload:
                return True

        elif self._matching == 1:
            if hashlib.sha256(verifieddata).digest() == self._payload:
                return True

        elif self._matching == 2:
            if hashlib.sha512(verifieddata).digest() == self._payload:
                return True

        else:
            # currently only 0, 1 and 2 are assigned
            logging.warning("Only matching types 0, 1 and 2 supported\n")

        return False



    @property
    def usage(self):
        """Usage for this TLSA record"""
        return self._usage


    @property
    def selector(self):
        """Selector for this record"""
        return self._selector


    @property
    def matching(self):
        """Way to match data against certificate"""
        return self._matching


    @property
    def payload(self):
        """Payload data of the TLSA record"""
        return self._payload


    def __repr__(self):
        hexencoder = codecs.getencoder('hex')
        return '<TLSA %d %d %d %s>' % (self._usage, self._selector, self._matching, hexencoder(self._payload)[0].decode())



def get_tlsa_records(resolver, name):
    """Extracts all TLSA records for a given name"""

    logging.debug("searching for TLSA record on %s", name)
    s, r = resolver.resolve(name, rrtype=RR_TYPE_TLSA)
    if 0 != s:
        ub_strerror(s)
        return

    if r.data is None:
        logging.warning("No TLSA record returned")
        return set()

    result = set()
    for record in r.data.data:
        usage = record[0]
        selector = record[1]
        matching = record[2]
        data = record[3:]
        result.add(TLSARecord(usage, selector, matching, data))
        
    return result


def match_tlsa_records(records, certificates):
    """Returns all TLSA records matching the certificate"""

    usedrecords = set()
    result = 0

    for certificate in certificates:
        recfound = False

        for record in records:
            if record.match(certificate):
                logging.info("Matched record %s", record)
                usedrecords.add(record)
                recfound = True

        if not recfound:
            logging.error("No TLSA record returned")
            result = 2

    for record in records:
        if not record in usedrecords:
            logging.warning("Unused record %s", record)
            if result == 0:
                result = 1

    return result


def verify_tlsa_record(resolver, record, certificate):
    logging.debug("searching for TLSA record on %s", record)
    s, r = resolver.resolve(record, rrtype=RR_TYPE_TLSA)
    if 0 != s:
        ub_strerror(s)
        return

    if r.data is None:
        logging.error("No TLSA record returned")
        return 2

    for record in r.data.data:
        hexencoder = codecs.getencoder('hex')
        usage = ord(record[0])
        selector = ord(record[1])
        matching = ord(record[2])
        data = record[3:]

        if usage != 3:
            logging.warning("Only 'Domain-issued certificate' records supported\n")

        if selector == 0:
            verifieddata = certificate
        elif selector == 1:
            verifieddata = get_spki(certificate)
        else:
            # currently only 0 and 1 are assigned
            sys.stderr.write("Only selectors 0 and 1 supported\n")

        if matching == 0:
            if verifieddata == data:
                logging.info("Found matching record: `TLSA %d %d %d %s`",
                             usage, selector, matching, hexencoder(data)[0])
                return 0
        elif matching == 1:
            if hashlib.sha256(verifieddata).digest() == data:
                logging.info("Found matching record: `TLSA %d %d %d %s`",
                             usage, selector, matching, hexencoder(data)[0].decode())
                return 0
        elif matching == 2:
            if hashlib.sha512(verifieddata).digest() == data:
                logging.info("Found matching record: `TLSA %d %d %d %s`",
                             usage, selector, matching, hexencoder(data)[0].decode())
                return 0
        else:
            # currently only 0, 1 and 2 are assigned
            logging.warning("Only matching types 0, 1 and 2 supported\n")

    logging.error("could not verify any tlsa record\n")
    return 2


check_dane/xmpp.py
#!/usr/bin/python3

#!/usr/bin/python3

from __future__ import print_function

import argparse
import logging

from ssl import SSLContext, PROTOCOL_TLSv1_2, CERT_REQUIRED
from socket import socket

from check_dane.tlsa import get_tlsa_records
from check_dane.cert import add_certificate_options
from check_dane.abstract import DaneChecker
from check_dane.resolve import Resolver, srv_lookup

XMPP_OPEN = ("<stream:stream xmlns='jabber:{0}' xmlns:stream='"
             "http://etherx.jabber.org/streams' xmlns:tls='http://www.ietf.org/rfc/"
             "rfc2595.txt' to='{1}' xml:lang='en' version='1.0'>")
XMPP_CLOSE = "</stream:stream>"
XMPP_STARTTLS = "<starttls xmlns='urn:ietf:params:xml:ns:xmpp-tls'/>"

class XmppDaneChecker(DaneChecker):
    def _init_connection(self, family, host, port):

        logging.debug("Connecting to %s:%d", host, port)

        connection = socket(family=family)
        connection.connect((host, port))

        connection.sendall(XMPP_OPEN.format(self.servicetype, self._hostname).encode())
        answer = connection.recv(4096)
        logging.debug(answer)

        if not b'</stream:features>' in answer:
            answer = connection.recv(4096)
            logging.debug(answer)

        connection.sendall(XMPP_STARTTLS.encode())
        answer = connection.recv(4096)
        logging.debug(answer)

        print(host, self._hostname)
        connection = self._sslcontext.wrap_socket(connection, server_hostname=self._hostname)
        connection.do_handshake()

        connection.sendall(XMPP_OPEN.format(self.servicetype, self._hostname).encode())
        answer = connection.recv(4096)
        logging.debug(answer)

        if not b'</stream:features>' in answer:
            answer = connection.recv(4096)
            logging.debug(answer)

        return connection


    @property
    def port(self):
        return self._port


    @property
    def servicetype(self):
        return self._type


    def _gather_certificates(self):
        result = set()
        for (host, port), meta in self._endpoints:
            self._host = host
            self._port = port
            self._type = meta['type']
            result.update(DaneChecker._gather_certificates(self))

        return result


    def _gather_records(self):
        result = set()
        for (host, port), _ in self._endpoints:
            print(repr((host, port)))
            result.update(get_tlsa_records(self._resolver, "_%d._tcp.%s" % (port, host)))

        return result


    def _close_connection(self, connection):
        connection.send(XMPP_CLOSE.encode())
        answer = connection.recv(512)
        logging.debug(answer)
        connection.close()


    def __init__(self):
        self._port = None
        self._ssl = None
        self._host = None
        self._hostname = None
        DaneChecker.__init__(self)


    def set_args(self, args):
        DaneChecker.set_args(self, args)

        sslcontext = SSLContext(PROTOCOL_TLSv1_2)
        sslcontext.verify_mode = CERT_REQUIRED
        sslcontext.load_verify_locations(args.castore)

        cresolver = Resolver(args.ancor)
        self._sslcontext = sslcontext

        self._hostname = args.Host.encode('idna').decode()
        endpoints = []
        if not args.s2s:
            for endpoint, meta in srv_lookup("_xmpp-client._tcp.%s" %
                                             self._hostname,
                                             cresolver):
                meta['type'] = 'client'
                endpoints.append((endpoint, meta))
        if not args.c2s:
            for endpoint, meta in srv_lookup("_xmpp-server._tcp.%s" %
                                             self._hostname,
                                             cresolver):
                meta['type'] = 'server'
                endpoints.append((endpoint, meta))

        self._endpoints = endpoints


    def generate_menu(self, argparser):
        DaneChecker.generate_menu(self, argparser)
        group = argparser.add_mutually_exclusive_group()
        group.add_argument("--s2s", action="store_true",
                           help="Only check server-to-server connections")
        group.add_argument("--c2s", action="store_true",
                           help="Only check client-to-server connections")
        argparser.add_argument("-p", "--port",
                               action="store", type=int, default=0,
                               help="SMTP port")




def main():
    logging.basicConfig(format='%(levelname)5s %(message)s')
    checker = XmppDaneChecker()
    parser = argparse.ArgumentParser()

    parser.add_argument("--verbose", action="store_true")
    parser.add_argument("--quiet", action="store_true")

    checker.generate_menu(parser)
    add_certificate_options(parser)

    args = parser.parse_args()
    checker.set_args(args)

    if args.verbose:
        logging.getLogger().setLevel(logging.DEBUG)
    elif args.quiet:
        logging.getLogger().setLevel(logging.WARNING)
    else:
        logging.getLogger().setLevel(logging.INFO)

    return checker.check()


if __name__ == '__main__':
    import sys
    sys.exit(main())


check_dane/https.py
#!/usr/bin/python3

from __future__ import print_function

import argparse
import logging

from ssl import SSLContext, PROTOCOL_TLSv1_2, CERT_REQUIRED
from socket import socket

from check_dane.cert import add_certificate_options
from check_dane.abstract import DaneChecker


class HttpsDaneChecker(DaneChecker):
    def _init_connection(self, family, host, port):
        connection = self._sslcontext.wrap_socket(socket(family),
                                                  server_hostname=host)
        connection.connect((host, port))
        connection.send(b"HEAD / HTTP/1.1\r\nHost: %s\r\n\r\n" % host.encode())
        answer = connection.recv(512)
        logging.debug(answer)

        return connection


    @property
    def port(self):
        return self._port


    def _close_connection(self, connection):
        connection.close()


    def __init__(self):
        DaneChecker.__init__(self)


    def set_args(self, args):
        DaneChecker.set_args(self, args)

        self._port = args.port

        sslcontext = SSLContext(PROTOCOL_TLSv1_2)
        sslcontext.verify_mode = CERT_REQUIRED
        sslcontext.load_verify_locations(args.castore)

        self._sslcontext = sslcontext


    def generate_menu(self, argparser):
        DaneChecker.generate_menu(self, argparser)
        argparser.add_argument("-p", "--port",
                               action="store", type=int, default=443,
                               help="HTTPS port")




def main():
    logging.basicConfig(format='%(levelname)5s %(message)s')
    checker = HttpsDaneChecker()
    parser = argparse.ArgumentParser()

    parser.add_argument("--verbose", action="store_true")
    parser.add_argument("--quiet", action="store_true")

    checker.generate_menu(parser)
    add_certificate_options(parser)

    args = parser.parse_args()
    checker.set_args(args)

    if args.verbose:
        logging.getLogger().setLevel(logging.DEBUG)
    elif args.quiet:
        logging.getLogger().setLevel(logging.WARNING)
    else:
        logging.getLogger().setLevel(logging.INFO)

    return checker.check()


if __name__ == '__main__':
    import sys
    sys.exit(main())


check_dane/resolve.py
#!/usr/bin/python3

import struct
import logging
from datetime import datetime

from unbound import ub_ctx, ub_strerror
from unbound import RR_TYPE_A, RR_TYPE_AAAA, RR_TYPE_RRSIG, RR_TYPE_SRV

from ldns import ldns_wire2pkt
from ldns import LDNS_SECTION_ANSWER


def _parse_rrsig_date(expirestring):
    expires = datetime(int(expirestring[:4]),
                       int(expirestring[4:6]),
                       int(expirestring[6:8]),
                       int(expirestring[8:10]),
                       int(expirestring[10:12]),
                       int(expirestring[12:14]))
    return expires


def format_address(data, datatype):
    """Given a answer packet for an A or AAAA query, return the string
       representation of the address
    """
    if datatype == RR_TYPE_A:
        return '.'.join([str(a) for a in data])
    elif datatype == RR_TYPE_AAAA:
        data = list(struct.iter_unpack("!H", data))
        return ":".join(["%x" % a for a in data])
    else:
        return None


def dnssec_verify_rrsig_validity(data, warn=-1, critical=0):
    """Given a answer packet confirm validity of rrsigs (with safety) """
    now = datetime.utcnow()

    s, packet = ldns_wire2pkt(data)
    if s != 0:
        logging.error("Parsing packet failed with errorcode %d", s)
        return 2

    rrsigs = packet.rr_list_by_type(RR_TYPE_RRSIG, LDNS_SECTION_ANSWER).rrs()
    rrsig = next(rrsigs)

    expire = _parse_rrsig_date(str(rrsig.rrsig_expiration()))
    incept = _parse_rrsig_date(str(rrsig.rrsig_inception()))

    if now < incept:
        logging.error("Signature not yet valid, only from %s", incept)
        return 2

    stillvalid = expire - now
    deltastr = str(stillvalid).split(",")

    if stillvalid.days < max(0, critical):
        logging.error("expires in %8s,%16s", deltastr[0], deltastr[1])
        return 2
    elif stillvalid.days < warn:
        logging.warning("expires in %8s,%16s", deltastr[0], deltastr[1])
        return 1


def srv_lookup(name, resolver):
    retval = []
    result = resolver.resolve(name, rrtype=RR_TYPE_SRV)
    for bytevalue in result.data.raw:
        priority, weight, port = struct.unpack("!HHH", bytevalue[:6])
        hostname = '.'.join(result.data.dname2str(bytevalue[6:]))
        retval.append(((hostname, port), {'priority': priority, 'weight': weight}))
    return retval


class ResolverException(BaseException):
    def __init__(self, message):
        BaseException.__init__(self)
        self.message = message


class Resolver:
    def __init__(self, ancor, fwd=None):
        self._resolver = ub_ctx()
        status = self._resolver.add_ta_file(ancor)
        if status != 0:
            raise ResolverException(ub_strerror(status))

        if fwd is not None:
            status = self._resolver.set_fwd(fwd)
            if status != 0:
                raise ResolverException(ub_strerror(status))


    def resolve(self, name, rrtype, secure=False):
        status, result = self._resolver.resolve(name, rrtype)
        if 0 != status:
            raise ResolverException(ub_strerror(status))

        if secure and not result.secure:
            raise ResolverException("Response was not signed")

        return result


check_dane/abstract.py
from abc import ABC, abstractmethod
from socket import AF_INET6, AF_INET

from unbound import ub_ctx

from check_dane.cert import verify_certificate
from check_dane.tlsa import get_tlsa_records, match_tlsa_records


class DaneWarning:
    pass

class DaneError:
    pass


class DaneChecker(ABC):
    def __init__(self):
        pass


    @abstractmethod
    def _init_connection(self, family, host, port):
        pass


    @abstractmethod
    def _close_connection(self, connection):
        pass


    @property
    @abstractmethod
    def port(self):
        pass


    def _gather_certificates(self):
        retval = 0
        certificates = set()
        for afamily in self._afamilies:
            connection = self._init_connection(afamily, self._host, self.port)

            nretval = verify_certificate(connection.getpeercert(), self._args)
            retval = max(retval, nretval)
            certificates.add(connection.getpeercert(binary_form=True))

            self._close_connection(connection)

        return certificates


    def _gather_records(self):
        return get_tlsa_records(self._resolver, "_%d._tcp.%s" % (self.port, self._host))


    def generate_menu(self, argparser):
        argparser.add_argument("Host")

        argparser.add_argument("--check-dane",
                               action="store_false",
                               help="Verify presented certificate via DANE (default: enabled)")
        argparser.add_argument("--check-ca",
                               action="store_false",
                               help="Verify presented certificate via the CA system (default: enabled)")
        argparser.add_argument("--check-expire",
                               action="store_false",
                               help="Verify presented certificate for expiration (default: enabled)")

        argparser.add_argument("-a", "--ancor",
                               action="store", type=str, default="/usr/share/dns/root.key",
                               help="DNSSEC root ancor")
        argparser.add_argument("--castore", action="store", type=str,
                               default="/etc/ssl/certs/ca-certificates.crt",
                               help="ca certificate bundle")

        group = argparser.add_mutually_exclusive_group()
        group.add_argument("-6", "--6", action="store_true", dest="use6", help="check via IPv6 only")
        group.add_argument("-4", "--4", action="store_true", dest="use4", help="check via IPv4 only")


    def set_args(self, args):
        self._args = args
        resolver = ub_ctx()
        resolver.add_ta_file(args.ancor)
        self._resolver = resolver

        if args.use6:
            self._afamilies = [AF_INET6]
        elif args.use4:
            self._afamilies = [AF_INET]
        else:
            self._afamilies = [AF_INET, AF_INET6]

        self._host = args.Host.encode('idna').decode()


    def check(self):
        records = self._gather_records()
        certificates = self._gather_certificates()
        return match_tlsa_records(records, certificates)


check_dane/smtp.py
#!/usr/bin/python3

#!/usr/bin/python3

from __future__ import print_function

import argparse
import logging

from ssl import SSLContext, PROTOCOL_TLSv1_2, CERT_REQUIRED
from socket import socket

from check_dane.cert import add_certificate_options
from check_dane.abstract import DaneChecker


class SmtpDaneChecker(DaneChecker):
    def _init_connection(self, family, host, port):

        if self.ssl:
            connection = self._sslcontext.wrap_socket(socket(family),
                                                      server_hostname=host)
            connection.connect((host, port))
            answer = connection.recv(512)
            logging.debug(answer)

            connection.send(b"EHLO localhost\r\n")
            answer = connection.recv(512)
            logging.debug(answer)

        else:
            connection = socket(family=family)
            connection.connect((host, port))
            answer = connection.recv(512)
            logging.debug(answer)

            connection.send(b"EHLO localhost\r\n")
            answer = connection.recv(512)
            logging.debug(answer)

            connection.send(b"STARTTLS\r\n")
            answer = connection.recv(512)
            logging.debug(answer)

            connection = self._sslcontext.wrap_socket(connection, server_hostname=host)
            connection.do_handshake()

            connection.send(b"EHLO localhost\r\n")
            answer = connection.recv(512)
            logging.debug(answer)

        return connection


    @property
    def port(self):
        return self._port


    @property
    def ssl(self):
        return self._ssl


    def _close_connection(self, connection):
        connection.send(b"QUIT\r\n")
        answer = connection.recv(512)
        logging.debug(answer)
        connection.close()


    def __init__(self):
        self._port = None
        self._ssl = None
        DaneChecker.__init__(self)


    def set_args(self, args):
        DaneChecker.set_args(self, args)

        self._ssl = args.ssl
        if args.port == 0:
            self._port = 465 if args.ssl else 25
        else:
            self._port = args.port

        sslcontext = SSLContext(PROTOCOL_TLSv1_2)
        sslcontext.verify_mode = CERT_REQUIRED
        sslcontext.load_verify_locations(args.castore)

        self._sslcontext = sslcontext


    def generate_menu(self, argparser):
        DaneChecker.generate_menu(self, argparser)
        argparser.add_argument("-p", "--port",
                               action="store", type=int, default=0,
                               help="SMTP port")
        argparser.add_argument("--ssl",
                               action="store_true",
                               help="Use direct TLS connection instead of starttls (default: disabled)")





def main():
    logging.basicConfig(format='%(levelname)5s %(message)s')
    checker = SmtpDaneChecker()
    parser = argparse.ArgumentParser()

    parser.add_argument("--verbose", action="store_true")
    parser.add_argument("--quiet", action="store_true")

    checker.generate_menu(parser)
    add_certificate_options(parser)

    args = parser.parse_args()
    checker.set_args(args)

    if args.verbose:
        logging.getLogger().setLevel(logging.DEBUG)
    elif args.quiet:
        logging.getLogger().setLevel(logging.WARNING)
    else:
        logging.getLogger().setLevel(logging.INFO)

    return checker.check()


if __name__ == '__main__':
    import sys
    sys.exit(main())
